### DISCUSSION TITLE
Measuring tools custom formating

### DISCUSSION THREAD

**Discussion by salah-walid** [iTwin/itwinjs-core#4817] [Q&A] [answered]
Hi ITwin Team,

Using measuring tools from @itwin/measure-tools-react i would like to have a special formatting for the distances and surfaces, for example i would like the following to show 2.43 km instead of 2431.3326 m
![image](https://user-images.githubusercontent.com/29980658/208144185-164dc070-5fed-4f28-ad71-99efe35b400d.png)

Similarly for the imperial system, formatting the inches and feets in a nice format

**Comment by pmconne** [comment]
You should be able to customize the formatting using [IModelApp.quantityFormatter](https://www.itwinjs.org/learning/frontend/quantityformatting/).

**Reply by salah-walid** [reply]
Is there a way to override the default conversion functions(the function that receives the unit and the distance and return the formatted distance), for more flexibility.

**Reply by pmconne** [reply]
What is the name of the function to which you refer?
What specifically are you attempting to achieve that - having consulted the documentation - you feel that `QuantityFormatter` does not permit?

**Reply by salah-walid** [reply]
For an example with a distance in the imperial system i would like it to be formated as such `{feet}' {inches}"`

**Reply by pmconne** [reply]
That seems like the sort of thing `QuantityFormatter` ought to be able to do. However I see that relevant APIs like [FormatProps](https://www.itwinjs.org/reference/core-quantity/quantity/formatprops/) are marked `@beta` and sparsely documented; and thus far I've found no tests or examples illustrating usage that produces output similar to what you want.

@ColinKerr and @calebmshafer are designated as owners of this code. Perhaps they can help.

**Reply by pmconne** [reply]
@ColinKerr please advise.

**Comment by pmconne** [comment, marked as answer]
@salah-walid are you using the [MeasureDistanceTool](https://www.itwinjs.org/reference/core-frontend/measure/measuredistancetool/) from @itwin/core-frontend? If so, the following produces the formatting you requested. (I took a guess that you want floating-point values expressed as fractions; if you want decimal instead, use `type: "Decimal"`).

```
    IModelApp.quantityFormatter.setOverrideFormat(QuantityType.Length, {
      type: "Fractional",
      formatTraits: ["KeepSingleZero", "ShowUnitLabel", "Use1000Separator" ],
      composite: {
        spacer: " ",
        units: [{ name: "Units.FT", label: "'" }, { name: "Units.IN", label: '"' }],
      },
    });
```

The documentation for [FormatProps](https://www.itwinjs.org/reference/core-quantity/quantity/formatprops/) (and the rest of @itwin/core-quantity) is basically non-existent, and the types don't help much - they're all just `string`s with no indication as to what specific strings might be accepted.

I tried the same solution in [a sandbox](https://www.itwinjs.org/sandboxes/pmconne/measure-tool-quantity-formatting) with less encouraging results -all of my formatting options (with the exception of switching to imperial units) are ignored. I suspect the sandbox may be using a different set of measurement tools which are not fully utilizing QuantityFormatter.

**Reply by salah-walid** [reply]
No currently using @itwin/measure-tools-react [viewer-components-react](https://github.com/iTwin/viewer-components-react), using this doesn't change the formating.

**Reply by pmconne** [reply]
Please update to the latest version of the package - apparently [that was fixed](https://github.com/iTwin/viewer-components-react/issues/361#issuecomment-1398563059).

**Reply by salah-walid** [reply]
Currently using 0.10.6 and it still not working, this is how i am setting it up:
```Typescript
await MeasureTools.startup();
    IModelApp.quantityFormatter.setOverrideFormat(QuantityType.Length, {
      type: "Fractional",
      formatTraits: ["KeepSingleZero", "ShowUnitLabel", "Use1000Separator" ],
      composite: {
        spacer: " ",
        units: [{ name: "Units.FT", label: "'" }, { name: "Units.IN", label: '"' }],
      },
    });
    MeasurementUIEvents.onMeasurementsChanged.addListener(this.handleMeasuringChanged);
```

This function is called on the `IModelApp.viewManager.onViewOpen` event, Still showing this measuring format:
![image](https://user-images.githubusercontent.com/29980658/214100361-ae967a49-a63f-4e65-bda6-b0a3ae49df30.png)

**Reply by aruniverse** [reply]
Try `QuantityType.LengthEngineering` instead

**Reply by pmconne** [reply]
I can confirm changing to `LengthEngineering` makes the formatting work in [my sandbox](https://www.itwinjs.org/sandbox/pmconne/measure-tool-quantity-formatting).
We need to clean up the API and docs for this core-quantity package. It's not very usable. It also does annoying things like throw exceptions if you pass it any inputs that don't meet its strict and undocumented requirements, such as
> `Uncaught (in promise) Error (35040): The Format QuantityTypeEnumValue-9 has a Composite with an invalid 'spacer' attribute. It should be an empty or one character string.`

which causes all of your formatting changes to be silently rejected.

**Reply by salah-walid** [reply]
Thank you @pmconne it finally worked, also a list of possible units name would be helpful, or have that property typed, with string literals
